In [6]:
sentence = 'the quick brown fox jumps over the lazy dog'

In [7]:
type(sentence)

str

In [8]:
words = sentence.split()
words

['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

In [10]:
['quick', 'brown', 'fox']
words[1:4]

['quick', 'brown', 'fox']

In [11]:
['quick', 'brown', 'fox', 'lazy', 'dog']
words[1:4] + words[-2:]

['quick', 'brown', 'fox', 'lazy', 'dog']

In [16]:
# ['jumps', 'over', ... 'dog', 'the', 'quick', ...]
# 
# new-object soln
words[-5:] + words[:4]

['dog', 'the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy']

In [15]:
# in-place modification solution
first4 = words[:4]
words[:4] = []
words.extend(first4)
words

['jumps', 'over', 'the', 'lazy', 'dog', 'the', 'quick', 'brown', 'fox']

## Strings

In [47]:
name = 'Alenxandar'
address = "1st cross, 2'nd main"
address = '1\'st cross, 2\'nd main'
speech = "Julius Cesar said \"Et tu brute\" "
speech = 'Julius Cesar said "Et tu brute" '
multiline = 'a\
very\
long string'
multiline = '''a
very
long
string'''
winpath = r'c:\windows\test\newfile.txt'     # raw string
email_patt = r'\w+@\w+\.\w+'

In [18]:
type(address)

str

In [28]:
print(winpath)

c:\windows\test\newfile.txt


In [32]:
name + ' ' + address

"Alenxandar 1'st cross, 2'nd main"

In [33]:
name * 3

'AlenxandarAlenxandarAlenxandar'

In [34]:
(name + ' ') * 3

'Alenxandar Alenxandar Alenxandar '

In [36]:
name[-4:]

'ndar'

In [38]:
'Alar'
name[:2] + name[-2:]

'Alar'

In [39]:
name.capitalize()

'Alenxandar'

In [40]:
name.upper()

'ALENXANDAR'

In [41]:
name.lower()

'alenxandar'

In [42]:
name.index('a')

5

In [43]:
name

'Alenxandar'

In [44]:
name.index('alex')

ValueError: substring not found

In [45]:
name.find('alex')

-1

In [46]:
name = 'abc'
name = 123

In [48]:
name.startswith('alex')

False

In [49]:
name.endswith('dar')

True

In [63]:
num = input('Enter number:')
if num.isdigit():
      print(int(num) ** 2)
      print('hello')
else:
    print('Number expected')

IndentationError: unindent does not match any outer indentation level (<tokenize>, line 5)

In [64]:
num ** 2

TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'

In [ ]:
'A' -> 65  ...
'a' -> 97 ...

In [65]:
2 ** 8

256

In [66]:
2 ** 16  

65536

### Encoding

- ASCII - 1 byte - 256 values
- unicode - 2 bytes - 65536 - utf-8, utf-16

In [67]:
names = '      amar  akbar   anthony     '

In [68]:
names.strip()

'amar  akbar   anthony'

In [69]:
names.split()

['amar', 'akbar', 'anthony']

In [71]:
names.split().upper()

AttributeError: 'list' object has no attribute 'upper'

In [72]:
words

['jumps', 'over', 'the', 'lazy', 'dog', 'the', 'quick', 'brown', 'fox']

In [ ]:
'jumps:over:the:...'

In [73]:
':'.join(words)

'jumps:over:the:lazy:dog:the:quick:brown:fox'

In [74]:
names

'      amar  akbar   anthony     '

In [76]:
'amar akbar anthony'
' '.join(names.split())

'amar akbar anthony'

### Exercises

In [77]:
first = 'alexandar'
middle = 'GRAHAM'
last = 'BeLL'

In [ ]:
'A.G.Bell'

In [ ]:
first = 'mohandas'
second = 'karamchand'
last = 'gandhi'

In [80]:
first.upper()[0]
first[0].upper()

'A'

In [83]:
first.upper()[0] + '.' +  middle.upper()[0] + '.' + last.capitalize()

'A.G.Bell'

In [84]:
'.'.join([first.upper()[0], middle.upper()[0], last.capitalize()])

'A.G.Bell'

In [88]:
' '.join([first, middle, last]).title()

'Alexandar Graham Bell'

In [86]:
names.title()

'      Amar  Akbar   Anthony     '

#### Query

- list.reverse() vs reversed(list)

In [89]:
nums = [1,2,3,4]

In [90]:
nums.reverse()

In [91]:
nums

[4, 3, 2, 1]

In [92]:
nums.sort()

In [93]:
nums

[1, 2, 3, 4]

In [95]:
for num in nums.reverse():
    print(num)

TypeError: 'NoneType' object is not iterable

In [96]:
nums.reverse()
for num in nums:
    print(num)

1
2
3
4


In [94]:
None

In [97]:
for num in reversed(nums):
    print(num)

4
3
2
1


In [98]:
for num in sorted(nums):
    print(num)

1
2
3
4


### Im/Mutable, copy-by-reference/value

In [99]:
first = 'alexandar'
first.upper()

'ALEXANDAR'

In [100]:
first

'alexandar'

In [102]:
FIRST = first.upper()

In [103]:
FIRST

'ALEXANDAR'

In [ ]:
first = first.upper()

In [ ]:
FIRST = first

In [104]:
nums = [1,2,3,4]
nums.append(5)

In [105]:
nums2 = nums         # copy by reference
nums2.append(6)

In [106]:
nums2

[1, 2, 3, 4, 5, 6]

In [107]:
nums

[1, 2, 3, 4, 5, 6]

In [111]:
nums3 = nums.copy()         # copy by value
nums3.append(7)

In [108]:
id(nums)

4457284288

In [109]:
id(nums2)

4457284288

In [112]:
id(nums3)

4457556992

In [113]:
nums == nums2       # eq by value
nums is nums2       # eq by object i.e. same object

True

In [114]:
import copy

### Dictionaries

    Key-value stores optimized for lookup

- Lookup use-case
- Dictionary data type
- 

In [116]:
items   = ['cookies', 'biscuits', 'chocolates', 'toffees']
prices  =   [123.45,    10.0,      250,          2.50]

In [126]:
item = input('Enter item for price:')
if item in items:
    idx = items.index(item)
    print(prices[idx])
else:
    print(item, 'not present')

Enter item for price: coffee


coffee not present


In [ ]:
[['cookies', 123.45],  ..

In [124]:
print(1 in [1,2,3,4])
print(5 in [1,2,3,4])

True
False


In [123]:
'A' in 'Alex'

True

In [132]:
ages = {'Sleepy': 101, 'Jumpy': 106,  'Sneezy': 99}
squares = {10: 100, 11: 121, 12: 144}
words_bylen = {3: 'the', 5: 'quick', 4: 'that'}
misc = {1: 'adfasd', 'asdfsd': 23, }

ages = {
    'Sleepy'   :    101, 
    'Jumpy'    :    106,  
    'Sneezy'   :    99,
    'Sleepy'   :    105
}

In [128]:
type(ages)

dict

In [133]:
ages

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99}

In [134]:
ages['Sleepy']

105

In [135]:
ages['Jughead']

KeyError: 'Jughead'

In [136]:
'Jughead' in ages

False

In [137]:
ages.keys()

dict_keys(['Sleepy', 'Jumpy', 'Sneezy'])

In [138]:
ages.values()

dict_values([105, 106, 99])

In [139]:
ages.items()

dict_items([('Sleepy', 105), ('Jumpy', 106), ('Sneezy', 99)])

In [140]:
t = ('Sleepy', 105)

In [141]:
type(t)

tuple

In [142]:
t[0]

'Sleepy'

In [144]:
for item in t:
    print(item)

Sleepy
105


In [145]:
ages['Jughead'] = 111

In [146]:
ages

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99, 'Jughead': 111}

In [147]:
ages['Jumpy'] += 1

In [149]:
squares = {}

for num in range(10):
    squares[num] = num**2

In [150]:
squares

{0: 0, 1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64, 9: 81}

In [174]:
names = ['Sleepy', 'Jumpy', 'Sneezy']
ages = [105, 106, 99]
bdays = ['1/1/1911',]

In [168]:
d = {'Sleepy': (105, '1/1/1911')}

In [170]:
d['Sleepy'][0]

105

In [171]:
d = {'Sleepy': {
        'age': 105, 
        'bday': '1/1/1911'
    }
}

In [173]:
d['Sleepy']['age']

105

In [157]:
zip(names, ages)        # Lazy object
list(zip(names, ages))  # Eager object

[('Sleepy', 105), ('Jumpy', 106), ('Sneezy', 99)]

In [155]:
range(10)          # Lazy object
list(range(10))    # Eager object

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

In [159]:
age_dict = {}

for item in zip(names, ages):     
    # print(item[0], item[1])
    age_dict[item[0]] = item[1]

In [160]:
age_dict

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99}

In [161]:
dict(zip(names, ages))

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99}

In [163]:
for name in names:
    for age in ages:
        print(name, age)

Sleepy 105
Sleepy 106
Sleepy 99
Jumpy 105
Jumpy 106
Jumpy 99
Sneezy 105
Sneezy 106
Sneezy 99


In [164]:
d = {}
d[names[0]] = ages[0]
d[names[1]] = ages[1]
d[names[2]] = ages[2]
d

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99}

In [166]:
# Emulation of indexed for loop
d = {}
for i in range(len(names)):
    d[names[i]] = ages[i]

In [167]:
d

{'Sleepy': 105, 'Jumpy': 106, 'Sneezy': 99}

#### Aside: Python interpreter implementations

- CPython    : C
- IronPython : C#
- JPython : Java

In [183]:
dict(zip(names, zip(ages, bdays)))

{'Sleepy': (105, '1/1/1911')}

# https://github.com/py-train/npci-hyd